File: go/food-chain/food_chain.go
package foodchain

import (
	"fmt"
	"strings"
)

const testVersion = 3

var verseData = []struct {
	animal      string
	description string
}{
	{"fly", "I don't know why she swallowed the fly. Perhaps she'll die."},
	{"spider", "It wriggled and jiggled and tickled inside her."},
	{"bird", "How absurd to swallow a bird!"},
	{"cat", "Imagine that, to swallow a cat!"},
	{"dog", "What a hog, to swallow a dog!"},
	{"goat", "Just opened her throat and swallowed a goat!"},
	{"cow", "I don't know how she swallowed a cow!"},
	{"horse", "She's dead, of course!"},
}

// Song returns the entire song.
func Song() string {
	return Verses(1, len(verseData))
}

// Verses returns `count` verses beginning with the verse specified by `number`.
func Verses(number int, count int) string {
	verses := make([]string, count)

	for i := range verses {
		verses[i] = Verse(number + i)
	}

	return strings.Join(verses, "\n\n")
}

// Verse returns the verse specified by `number`.
func Verse(number int) string {
	stanzas := []string{}

	for i, n := 0, number-1; i < number; i, n = i+1, n-1 {
		if i == 0 {
			stanzas = append(
				stanzas,
				fmt.Sprintf("I know an old lady who swallowed a %s.", verseData[n].animal),
				verseData[n].description,
			)

			if number == len(verseData) {
				break
			}
		} else if n == 1 {
			stanza := fmt.Sprintf(
				"She swallowed the %s to catch the %s that %s",
				verseData[n+1].animal,
				verseData[n].animal,
				verseData[n].description[3:],
			)
			stanzas = append(stanzas, stanza)
		} else {
			stanzas = append(
				stanzas,
				fmt.Sprintf("She swallowed the %s to catch the %s.", verseData[n+1].animal, verseData[n].animal),
			)

			if n == 0 {
				stanzas = append(stanzas, verseData[n].description)
			}
		}
	}

	return strings.Join(stanzas, "\n")
}

File: ruby/food-chain/food_chain.rb
# frozen_string_literal: false

module FoodChain
  OLD_LADY = 'I know an old lady who swallowed a %{animal}.'
  REASON = 'She swallowed the %{animal} to catch the %{prey}%{details}.'

  ANIMALS = [
    'fly',
    'spider',
    'bird',
    'cat',
    'dog',
    'goat',
    'cow',
    'horse',
  ]

  REASONS = [
    'I don\'t know why she swallowed the fly. Perhaps she\'ll die.',
    'It wriggled and jiggled and tickled inside her.',
    'How absurd to swallow a bird!',
    'Imagine that, to swallow a cat!',
    'What a hog, to swallow a dog!',
    'Just opened her throat and swallowed a goat!',
    'I don\'t know how she swallowed a cow!',
    'She\'s dead, of course!',
  ]

  DETAILS = [
    '',
    ' that wriggled and jiggled and tickled inside her',
    '',
    '',
    '',
    '',
    '',
    '',
  ]

  def self.song
    1.upto(ANIMALS.size).map { |i| verse(i) }.join("\n")
  end

  def self.verse(number)
    lyrics = []

    lyrics << format(OLD_LADY, animal: ANIMALS[number - 1])
    lyrics << REASONS[number - 1] if number > 1

    if number < ANIMALS.size
      (number - 1).downto(1) do |meal|
        lyrics << format(REASON, animal: ANIMALS[meal], prey: ANIMALS[meal - 1], details: DETAILS[meal - 1])
      end

      lyrics << REASONS[0]
    end

    lyrics << ''

    lyrics.join("\n")
  end
end

module BookKeeping
  VERSION = 2
end
